language/vm/src/file_format.rs

```diff
@@ -1605,6 +1606,8 @@ pub struct CompiledScript(CompiledScriptMut);
 /// the bounds checker.
 #[derive(Clone, Default, Eq, PartialEq, Debug)]
 pub struct CompiledScriptMut {
+    /// Version number found during deserialization
+    pub version: u32,
```

Comment:
I feel like it might not be a good idea to include the version number as part of the `CompiledScript`. In our current model, the Rust representation doesn't really have a concept of versioning -- only the binary format does. If you include the version number, then `serialize(deserialize(script))` will no longer be an identity function. I'd recommend you make the deserializer return a tuple (`(u32, CompiledScript)`) instead.

Reply:
Yeah that's not really doable. And `serialize(deserialize(script))` already isn't an identity function because of the different versions. We always serialize into the most recent version

Reply:
My bad. I meant "deserialize(serialize(script))".

Reply:
That will still work, the version number isn't used at all in serialization

Reply:
Sorry for the late entry. I am a bit uncomfortable with having a version number in `CompiledScript` and `CompiledModule` too, especially when this number is not re-serialized into anything...

I like the tuple idea better, i.e., the deserializer return a tuple `(u32, CompiledScript)`, where the `u32` represents the version number of the bytecode when that module was originally serialized. With enough care, I think we might be able to confine the major changes in `move-vm` only (or more optimistically, the `loader.rs` only).

Reply:
> It will still deserialize but there is no guarantee you will get the same version number before. My point here really is that the concept of versioning does not currently belong to `CompiledScript` -- you'll only have this information if you create your `CompiledScript` from deserialization, but in theory it is possible for you to create a `CompiledScript` entirely in-memory and in that case you don't have a version number.

Wanted to say this, but on a second thought, I feel like maybe what you did isn't actually a bad idea... The correctness of a script is now a variable of the version number and if you want the script to be self-contained you have to include the version info, in one way or another. Anyway, need to think more about the implications of this. The PR is good to go in its current shape!